src/flag_gems/ops/flash_kernel.py

```diff
@@ -0,0 +1,1334 @@
+    n_splits,
+    BLOCK_M: tl.constexpr,
+    q_total,
+    MAX_N_SPLITS: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    lse_splits_ptr += pid * BLOCK_M
+    lse_ptr += pid * BLOCK_M
+    out_splits_ptr += pid * BLOCK_M * head_size
+    out_ptr += pid * BLOCK_M * head_size
+    lse_split_stride = tl.num_programs(0) * BLOCK_M
+    out_split_stride = tl.num_programs(0) * BLOCK_M * head_size
+
+    # Subtracting maximum from each of the split lse's for better numerical stability
+    lse_split_offset = (
+        tl.arange(0, BLOCK_M)[:, None]
+        + tl.arange(0, MAX_N_SPLITS)[None, :] * lse_split_stride
+    )
+    lse_split_mask = (pid * BLOCK_M + tl.arange(0, BLOCK_M)[:, None] < q_total) & (
+        tl.arange(0, MAX_N_SPLITS)[None, :] < n_splits
+    )
+    lse_splits = tl.load(
+        lse_splits_ptr + lse_split_offset, mask=lse_split_mask, other=float("-inf")
+    )
+    max_lse = tl.max(lse_splits, 1)
+
+    # Sum exp(lse(i) - max_lse) over all split i to obtain Z=sumexp(QK) up to a scaled factor exp(-max_lse)
+    Zi_scaled = tl.exp(lse_splits - max_lse[:, None])
+    Z_scaled = tl.sum(Zi_scaled, 1)
+    Zi_Z = Zi_scaled / Z_scaled[:, None]
+
+    # Write back LSE
+    lse = tl.log(Z_scaled) + max_lse
+    out_mask = pid * BLOCK_M + tl.arange(0, BLOCK_M) < q_total
+    tl.store(lse_ptr + tl.arange(0, BLOCK_M), lse, mask=out_mask)
+
+    out_split_offset = (
+        tl.arange(0, BLOCK_M)[:, None, None] * head_size
+        + tl.arange(0, MAX_N_SPLITS)[None, :, None] * out_split_stride
+        + tl.arange(0, head_size)[None, None, :]
+    )
+    out_split_mask = (
+        pid * BLOCK_M + tl.arange(0, BLOCK_M)[:, None, None] < q_total
+    ) & (tl.arange(0, MAX_N_SPLITS)[None, :, None] < n_splits)
+    out_splits = tl.load(
+        out_splits_ptr + out_split_offset, mask=out_split_mask, other=0
+    )
+    out = tl.sum(Zi_Z[:, :, None] * out_splits, 1)
+    out = out.to(out_ptr.type.element_ty)
+
+    # Write back output
+    out_offset = tl.arange(0, BLOCK_M)[:, None] * out_s_stride + tl.arange(0, head_size)
+    tl.store(out_ptr + out_offset, out, mask=out_mask[:, None])
+
+
+@triton.jit
+def block_to_cache_index(
+    n_block, page_table_ptr, block_size, page_stride, row_stride, BLOCK_N
+):
+    row_index = n_block * BLOCK_N
+    page_offset = row_index % block_size
+    virtual_page_index = row_index // block_size
+    # page_table_ptr is already pointed at the start of the current batch element
+    cache_page_index = tl.load(page_table_ptr + virtual_page_index).to(tl.int32)
+    return cache_page_index * block_size + page_offset
+
+
+@triton.jit
+def flash_varlen_fwd_kernel(
+    q_ptr,
+    k_ptr,
+    v_ptr,
+    o_ptr,
+    p_ptr,
+    softmax_lse_ptr,
+    q_row_stride,
+    k_row_stride,
+    v_row_stride,
+    q_head_stride,
+    k_head_stride,
+    v_head_stride,
+    o_row_stride,
+    o_head_stride,
+    q_batch_stride,
+    k_batch_stride,
+    v_batch_stride,
+    o_batch_stride,
+    is_cu_seqlens_q,
```

Comment:
There is `if is_cu_seqlens_q` in the code. Should `is_cu_seqlens_q` be marked as tl.constexpr?